### PR TITLE
Add evdev dispatcher pointer + keyboard

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by neithernut** [comment]
ACK for squashing (splittign the last commit and integrating it into the other ones).

File: src/input/input_device.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    ev_io_start(default_loop, &self->watcher);
+
+    return self;
+}
+
+static size_t
+hash_callback(
+    struct ws_object * _self
```
Comment (matthiasbeyer) [inline src/input/input_device.c]: I guess you do not need the underscore in the variable name here, do you?

File: src/input/input_device.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+static int
+cmp_callback(
+    struct ws_object const* obj1,
+    struct ws_object const* obj2
+) {
+    struct ws_input_device* dev1 = (struct ws_input_device*) obj1;
+    struct ws_input_device* dev2 = (struct ws_input_device*) obj2;
+    // It's short signum function
+    return (dev1->fd > dev2->fd) - (dev1->fd < dev2->fd);
```
Comment (matthiasbeyer) [inline src/input/input_device.c]: Why don't you use our macro for this?

File: src/input/utils.c [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ws_input_filter_event_device_name(
+    const char* name
+) {
+    if (!(strcmp(".", name) && strcmp("..", name))) {
+        return false;
+    }
+    // strcmp also makes note of the length between the strings, we only want
+    // it to check the beginning of the file name, in this case it has to match
+    // "event*"
+    if (strncmp(INPUT_DEVICE_NAME, name, sizeof(INPUT_DEVICE_NAME)-1)) {
```
Comment (matthiasbeyer) [inline src/input/utils.c]: Why the `-1` ?

Comment (neithernut) [inline src/input/utils.c]: To omit the `\0`

File: src/compositor/cursor.c [hunk 4]
```diff
@@ -104,6 +104,11 @@ ws_cursor_set_position(
     // off screen.
     self->x = CLAMP(-self->x_hp, x, w);
     self->y = CLAMP(-self->y_hp, y, h);
+    int retval = drmModeMoveCursor(self->cur_fb_dev->fd, self->cur_mon->crtc,
+            self->x, self->y);
+    if ( retval != 0) {
```
Comment (matthiasbeyer) [inline src/compositor/cursor.c]: trailing whitespace here.

File: src/input/input_device.c [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        if (retval != LIBEVDEV_READ_STATUS_SUCCESS) {
+            ws_log(&log_ctx, LOG_ERR, "libevdev unspecified error %d", retval);
+            continue;
+        }
+
+        if (ev.type != EV_REL) {
+            continue;
+        }
+
+        //ws_log(&log_ctx, LOG_DEBUG, "It's a mouse event!");
```
Comment (matthiasbeyer) [inline src/input/input_device.c:174]: Outcommented because it generates a lot output, right?

Comment (TheNeikos) [inline src/input/input_device.c:174]: Yes, it's meant to be there for easy debugging should it be needed.

File: src/input/utils.c (hunk 3 above)
Comment (matthiasbeyer) [inline src/input/utils.c]: Missing whitespace around `-` here.

File: src/input/input_device.c (hunk 5 above)
Comment (TheNeikos) [inline src/input/input_device.c:174]: To give an order of magnitude, this generates about 250 MB in 10 seconds on my machine

Comment (matthiasbeyer) [inline src/input/input_device.c:174]: But when it is for debugging, you can control it through a env variable, so it shouldn't be a problem...

File: src/compositor/cursor.c [hunk 6]
```diff
@@ -104,6 +104,11 @@ ws_cursor_set_position(
     // off screen.
     self->x = CLAMP(-self->x_hp, x, w);
     self->y = CLAMP(-self->y_hp, y, h);
+    int retval = drmModeMoveCursor(self->cur_fb_dev->fd, self->cur_mon->crtc,
+            self->x, self->y);
```
Comment (matthiasbeyer) [inline src/compositor/cursor.c]: Indentation!

File: src/input/input_device.c (hunk 5 above)
Comment (TheNeikos) [inline src/input/input_device.c:174]: No. This is even under `LOG_DEBUG` it's on the `LOG_LAST_RESORT` level.

Comment (matthiasbeyer) [inline src/input/input_device.c:174]: :smile:

Comment (matthiasbeyer) [inline src/input/input_device.c:174]: lastresort

**Comment by neithernut** [comment]
ACK.

**Comment by matthiasbeyer** [comment]
ACK.

**Pull request merged**